### ISSUE TITLE
Change listener not being triggered on realm updates

### ISSUE THREAD

**Issue by Wattos** [realm/realm-java#1676]
According to the documentation:

*"If you want to be notified when the query completes and the RealmResults is updated, you can register a RealmChangeListener. This listener will be called every time the RealmResults is updated to reflect the latest changes in the Realm (usually after a commit)"*

I add a change listener to either a RealmObject or RealmResult, but that only gets triggered once (after the data is loaded). The documentation states that this listener should also be triggered when the real result is updated. That does not seem to be the case. 

Is the documentation incorrect or do I need to do something special besides commiting a transaction so that the listener gets triggered?

**Comment by cmelchior** [comment]
Hi @Wattos 
The ChangeListener should be triggered on each update, not only when the data is loaded. If this isn't happening it sounds like a bug.

**Comment by nhachicha** [comment]
Hi @Wattos 
It would be helpful if you can send us some code to reproduce this (either here or privately to help@realm.io)
Cheers,

**Comment by Wattos** [comment]
@nhachicha Do you think that the code is really necessary? I did a quick search in the realm-java repository and the notifyChangeListener method is only called 3 times:

**Comment by vikas-fueled** [comment]
I have encountered similar issue with when at times change listener is not being triggered even once for the same query, but is triggered when I query it again. `isLoaded` returns false in both cases on checking on same thread after registering listener.
```
public void fetchEpisodes(long seasonId) {
        logger.debug("fetch episodes for Season id :" + seasonId);
        final Realm realm = Realm.getDefaultInstance();
        RealmQuery<Content> query = realm.where(Content.class)
                .equalTo("season.id", seasonId);
        final RealmResults<Content> contents = query
                .findAllAsync();
        contents.addChangeListener(new RealmChangeListener() {
            @Override
            public void onChange() {
                logger.debug("Episodes size :" + contents.size());
                if (adapter != null) {
                    adapter.setEpisodes(contents, coverHandlerProvider);
                }
            }
        });
        logger.debug("Sync Episodes size :" + contents.size() + ", is Loaded :" + contents.isLoaded());
    }
```

**Comment by vikas-fueled** [comment]
From [this discussion](https://github.com/realm/realm-java/issues/1675) seems like anonymous listeners should be fine for `RealmResults`. However when I changed my listener to a concrete one, the above issue was solved. So, this might help in investigating further.

**Comment by cmelchior** [comment]
@Wattos I tried to reproduce your case, and it looks like it is working as intended. If you can provide some sample code where this is not happening it would be really helpfull.

**Comment by fuwaneko** [comment]
I have the similar problem and I'm pretty sure it happens *only* with queries on related models, i.e. .equalsTo("related.id", id). RealmResults is never updated. Listener actually *fires* twice (first immediately after findAllAsync I guess, second time after data has been written to Realm from network thread), but no data is present in RealmResults! After something else updates Realm listener fires again and data appears. But, if I remove condition on related model, everything works as it should.

**Comment by cmelchior** [comment]
@Wattos I discovered that you are right. It seems that our change listeners doesn't get triggered for our synchronous queries and objects. I have updated #1696 with the unit tests to catch this. /cc @nhachicha

**Comment by fuwaneko** [comment]
@cmelchior I've created test app to showcase my problems: https://dl.dropboxusercontent.com/u/11334864/master.zip

**Comment by nhachicha** [comment]
fixed in `0.87.1`

**Comment by jemshit** [comment]
This issue still exists in Version (`io.realm:realm-gradle-plugin:1.1.0`) and probably in latest version too. I can reproduce issue, but it is a bit complex. 

I have this method which is called by background thread by this [lib](https://github.com/yigit/android-priority-jobqueue):

```java
public void saveImageDownloaded(boolean isDownloaded){
        RealmManager realmManager = null;
        try{
            realmManager=new RealmManager();
            realmManager.updateImageDownloaded(isDownloaded,imageDatabaseType, imageName);
        }finally {
            if(realmManager!=null)realmManager.close();
        }
    }
```
RealmManager.java:

```java
...
realm=Realm.getDefaultInstance();
...
public void updateImageDownloaded(...) {
    realm.executeTransactionAsync(new Realm.Transaction() {
                    @Override
                    public void execute(Realm bgRealm) {
                        ModelRealm_ImageDestination model = bgRealm.where(ModelRealm_ImageDestination.class)
                                .equalTo("imageName", imageName)
                                .findFirst();
                        if (model != null)
                            model.setDownloaded(isDownloaded);
                    }
                });
}
...
public void close() {        
        if (realm != null) {
            realm.close();
        }
}
```
Problem **occurs** in exact above calling order, which is not quite right, because i run async call in `try` block and call `realm.close()` right after in `finally` block.

Problem **does not** occur in this case:
```java
public void saveImageDownloaded(boolean isDownloaded){
        RealmManager realmManager = null;
        try{
            realmManager=new RealmManager();
            realmManager.updateImageDownloaded(isDownloaded,imageDatabaseType, imageName);
        }finally {
            ///if(realmManager!=null)realmManager.close();
        }
    }
``` 
Problem **does not** occur in this case
```java
public void saveImageDownloaded(boolean isDownloaded){
        RealmManager realmManager = null;
        try{
            realmManager=new RealmManager();
            //realmManager.updateImageDownloaded(isDownloaded,imageDatabaseType, imageName);
        }finally {
            if(realmManager!=null)realmManager.close();
        }
    }
```

**Comment by Zhuinden** [comment]
You should consider closing the Realm instance only AFTER the transaction is completed 

http://stackoverflow.com/a/35402728/2413303

**Comment by jemshit** [comment]
I am aware of close() is called too early, so it should give error like "realm cannot be used after realm instance is closed.." if async op tries to do sth after .close(). But this above code breaks my RealmManager in every screen, even if i create new realm instance and try to use it.

**Comment by Zhuinden** [comment]
Define "breaks whole app until restart"

**Comment by jemshit** [comment]
realm.getDefaultInstance() does not work in any screen, for .findAllAsync(), changeListener does not get notified

**Comment by Zhuinden** [comment]
@jemshit I can imagine that occurring if you access the Realm from multiple processes. 

By the way, you should consider updating to 1.1.1, it's much more stable than 1.1.0

**Comment by GitHubMurt** [comment]
_classpath "io.realm:realm-gradle-plugin:3.5.0"_

private val listener = RealmChangeListener<T> { t ->
        if (t != null && t.isValid)
            realm.executeTransaction {
                value = realm.copyFromRealm(t)
            }

    }

Listener isn't called after saveCarDetails
```
fun saveCarDetails(carDetails: CarDetails){
        mRealm.executeTransactionAsync { realm ->
            carDetails.id = 1 // keep only one copy of the object
            realm?.copyToRealmOrUpdate(carDetails)
        }
    }

fun getCarDetails(): LiveData<CarDetails> {
        return (mRealm.where(CarDetails::class.java).equalTo(CarDetails.ID, 1).findFirstAsync()).asLiveData(mRealm)
    }
```

**Comment by cmelchior** [comment]
@GitHubMurt Please create a new issue with all details.

**Comment by Zhuinden** [comment]
@GitHubMurt you did not include `asLiveData`

**Comment by GitHubMurt** [comment]
I've created new issue as cmelchior suggested
https://github.com/realm/realm-java/issues/4984
